syncmaster/worker/handlers/file/base.py

```diff
@@ -59,14 +69,61 @@ def read(self) -> DataFrame:
         return df
 
     def write(self, df: DataFrame) -> None:
-        writer = FileDFWriter(
-            connection=self.connection,
-            format=self.transfer_dto.file_format,
-            target_path=self.transfer_dto.directory_path,
-            options=self.transfer_dto.options,
+        tmp_path = os.path.join(self.transfer_dto.directory_path, ".tmp", str(self.run_dto.id))
+        try:
+            writer = FileDFWriter(
+                connection=self.df_connection,
+                format=self.transfer_dto.file_format,
+                target_path=tmp_path,
+                options=self.transfer_dto.options,
+            )
+            writer.run(df=df)
+
+            self._rename_files(tmp_path)
+
+            mover = FileMover(
+                connection=self.file_connection,
+                source_path=tmp_path,
+                target_path=self.transfer_dto.directory_path,
+                filters=[Glob(f"*.{self._get_file_extension()}")],
+            )
+            mover.run()
+        finally:
+            self.file_connection.remove_dir(tmp_path, recursive=True)
+
+    def _rename_files(self, tmp_path: str) -> None:
+        files = self.file_connection.list_dir(tmp_path)
+
+        for index, file_name in enumerate(files):
+            extension = self._get_file_extension()
+            new_name = self._get_file_name(str(index), extension)
+            old_path = os.path.join(tmp_path, file_name)
+            new_path = os.path.join(tmp_path, new_name)
+            self.file_connection.rename_file(old_path, new_path)
+
+    def _get_file_name(self, index: str, extension: str) -> str:
+        return self.transfer_dto.file_name_template.format(
+            index=index,
+            extension=extension,
+            run_id=self.run_dto.id,
+            run_created_at=self.run_dto.created_at.strftime("%Y_%m_%d_%H_%M_%S"),
         )
 
-        return writer.run(df=df)
+    def _get_file_extension(self) -> str:
+        extension = self.transfer_dto.file_format.name
+        compression = getattr(self.transfer_dto.file_format, "compression", None)
+        if not compression or compression == "none":
+            return extension
+
+        compression = self._compressions[compression]
+
+        if extension == "parquet" and compression == "lz4":
+            return "lz4hadoop.parquet"
+
+        if extension in ("parquet", "orc"):
```

Comment:
nit: we can add a constant like `COLUMN_FORMATS = {"parquet", "orc"}`, and use it in check